lookbook/docs/patterns/14-forms.md.erb

```diff
@@ -0,0 +1,228 @@
+Forms are used heavily in administration pages and throughout the application.
+They rely on Primer forms with some OpenProject specific components and helpers.
+
+## Overview
+
+<%= embed Primer::FormsPreview, :custom_width_fields_form, log: true %>
+
+## Usage
+
+To create forms, you need 2 basic things:
+
+- A form instance to render fields
+- A `primer_form_with` call to get a form builder and render the form instance
+
+```ruby
+class TextFieldAndCheckboxForm < ApplicationForm
+  form do |my_form|
+    my_form.text_field(
+      name: :ultimate_answer,
+      label: "Ultimate answer",
+      required: true,
+      caption: "The answer to life, the universe, and everything"
+    )
+
+    my_form.check_box(
+      name: :enable_ipd,
+      label: "Enable the Infinite Improbability Drive",
+      caption: "Cross interstellar distances in a mere nothingth of a second."
+    )
+  end
+end
+```
+
+```erb
+<%%= primer_form_with(url: "/foo") do |f| %>
+  <%%= render(TextFieldAndCheckboxForm.new(f)) %>
+<%% end %>
+```
+
+Multiple form instances can be rendered within the same `primer_form_with` call,
+allowing to put some content in between:
+
+```erb
+<%%= primer_form_with(url: "/foo") do |f| %>
+  <%%= render(TextFieldAndCheckboxForm.new(f)) %>
+  <%%= render(MessageComponent.new(icon: :info, message: "This will be fine!")) %>
+  <%%= render(SubmitButtonForm.new(f)) %>
+<%% end %>
+```
+
+This is the regular way of using Primer forms.
+
+## OpenProject helpers
+
+OpenProject provides some helpers to make building and rendering forms easier.
+
+### `render_inline_form` to avoid creating form classes
+
+This helper allows to render an anymous form instance, avoiding the need to
+create a dedicated form class. This can be useful for simple forms or when you
+don't want to pollute the form class namespace.
+
+```ruby
+class TextFieldWithWarningForm < ApplicationForm
+  attr_reader :warning
+
+  def initialize(warning: nil)
+    super()
+    @warning = warning
+  end
+
+  form do |my_form|
+    my_form.text_field(
+      name: :answer,
+      label: "answer",
+      required: true
+    )
+
+    if warning
+      my_form.html_content do
+        tag.div(class: "flash flash-warn") { warning }
+      end
+    end
+
+    my_form.submit(name: :submit, label: "Save")
+  end
+end
+```
```

Comment:
This example isn't really showing a use of a `render_inline_form`, I'd say it's demonstrating more the use of `FormObject#html_content` within a form class.

Reply:
Good point! I'll fix it.